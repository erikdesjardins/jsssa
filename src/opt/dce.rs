use crate::ir;
use crate::ir::fold::Folder;

#[derive(Default)]
pub struct Dce {
    dropping_after_jump: bool,
}

impl Folder for Dce {
    type Output = Option<ir::Stmt>;

    fn wrap_scope<R>(&mut self, enter: impl FnOnce(&mut Self) -> R) -> R {
        let r = enter(self);
        // stop dropping when we leave the current scope
        self.dropping_after_jump = false;
        r
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        if self.dropping_after_jump {
            return None;
        }

        match stmt {
            ir::Stmt::Expr {
                ref target,
                ref expr,
            } if target.used().is_never() => match expr {
                ir::Expr::Function { .. }
                | ir::Expr::Bool { .. }
                | ir::Expr::Number { .. }
                | ir::Expr::String { .. }
                | ir::Expr::Null
                | ir::Expr::Undefined
                | ir::Expr::This
                | ir::Expr::Read { .. }
                | ir::Expr::ReadMutable { .. }
                | ir::Expr::Array { .. }
                | ir::Expr::Object { .. }
                | ir::Expr::RegExp { .. }
                | ir::Expr::Unary { .. }
                | ir::Expr::Binary { .. }
                | ir::Expr::CurrentFunction
                | ir::Expr::Argument { .. } => None,
                ir::Expr::ReadGlobal { .. }
                | ir::Expr::ReadMember { .. }
                | ir::Expr::Delete { .. }
                | ir::Expr::Yield { .. }
                | ir::Expr::Await { .. }
                | ir::Expr::Call { .. } => Some(stmt),
            },
            ir::Stmt::DeclareMutable { ref target, val: _ } if target.used().is_never() => None,
            ir::Stmt::Return { .. }
            | ir::Stmt::Throw { .. }
            | ir::Stmt::Break
            | ir::Stmt::Continue => {
                self.dropping_after_jump = true;
                Some(stmt)
            }
            ir::Stmt::IfElse {
                cond: _,
                ref cons,
                ref alt,
            } if cons.0.is_empty() && alt.0.is_empty() => None,
            ir::Stmt::Try {
                ref body,
                catch: _,
                ref finally,
            } if body.0.is_empty() && finally.0.is_empty() => None,
            _ => Some(stmt),
        }
    }
}
