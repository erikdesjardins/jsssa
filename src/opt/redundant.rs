use std::collections::{HashMap, HashSet};
use std::mem;

use crate::anal;
use crate::ir;
use crate::ir::traverse::{Folder, RunVisitor, ScopeTy, Visitor};

/// Remove or forward redundant loads and stores of mutable vars.
///
/// Does not profit from multiple passes.
/// Does not profit from DCE running first; may create opportunities for DCE.
/// May create opportunities for mut-to-ssa downleveling.
/// May create opportunities for read forwarding.
///
/// Performs the following transformations:
///
/// Read-to-read forwarding:
///   something :- <null>
///   /* stuff */
///   x = something
///   y = something
/// ->
///   something :- <null>
///   /* stuff */
///   x = something
///   y = x
///
/// Write-to-read forwarding:
///   something :- <null>
///   /* stuff */
///   x = 1 + 1
///   something <- x
///   y = something
/// ->
///   something :- <null>
///   /* stuff */
///   x = 1 + 1
///   something <- x
///   y = x
///
/// Dead write elimination:
///   something :- <null>
///   /* stuff */
///   something <- x
///   something <- y
/// ->
///   something :- <null>
///   /* stuff */
///   something <- y
///
#[derive(Debug, Default)]
pub struct LoadStore {
    mut_ops_to_replace: HashMap<StmtIndex, What>,
    cur_index: StmtIndex,
}

type StmtIndex = u64;

#[derive(Debug)]
enum What {
    ReadSsa(ir::Ref<ir::Ssa>),
    Remove,
    BecomeDecl,
}

#[derive(Debug, Default)]
struct CollectLoadStoreInfo<'a> {
    refs_used_in_only_one_fn_scope: HashSet<&'a ir::Ref<ir::Mut>>,
    mut_ops_to_replace: HashMap<StmtIndex, What>,
    cur_index: StmtIndex,
    for_reads: Ops<'a>,
    for_writes: Ops<'a>,
}

#[derive(Debug, Default)]
struct Ops<'a> {
    local_last_op: HashMap<&'a ir::Ref<ir::Mut>, (StmtIndex, WriteOp<'a>)>,
    nonlocal_last_op: HashMap<&'a ir::Ref<ir::Mut>, (StmtIndex, WriteOp<'a>)>,
    invalid_in_parent: Invalid<'a>,
}

#[derive(Debug, Clone)]
enum WriteOp<'a> {
    Declare(&'a ir::Ref<ir::Ssa>),
    Write(&'a ir::Ref<ir::Ssa>),
}

// todo opt: change this to enum Everything, NonlocalAnd(HashSet), Refs(HashSet)
#[derive(Debug, Default)]
struct Invalid<'a> {
    all_local: bool,
    all_nonlocal: bool,
    refs: HashSet<&'a ir::Ref<ir::Mut>>,
}

impl<'a> CollectLoadStoreInfo<'a> {
    fn invalidate_from_child(&mut self, child: Self) {
        fn invalidate<'a>(this: &mut Ops<'a>, invalid: Invalid<'a>) {
            if invalid.all_local {
                this.local_last_op.clear();
                this.invalid_in_parent.all_local = true;
            }
            if invalid.all_nonlocal {
                this.nonlocal_last_op.clear();
                this.invalid_in_parent.all_nonlocal = true;
            }
            for invalid_ref in invalid.refs {
                this.local_last_op
                    .remove(&invalid_ref)
                    .or_else(|| this.nonlocal_last_op.remove(&invalid_ref));
                this.invalid_in_parent.refs.insert(invalid_ref);
            }
        }
        invalidate(&mut self.for_reads, child.for_reads.invalid_in_parent);
        invalidate(&mut self.for_writes, child.for_writes.invalid_in_parent);
    }

    fn invalidate_everything_used_across_fn_scopes(&mut self) {
        fn invalidate(this: &mut Ops<'_>) {
            this.nonlocal_last_op.clear();
            this.invalid_in_parent.all_nonlocal = true;
        }
        invalidate(&mut self.for_reads);
        invalidate(&mut self.for_writes);
    }

    fn invalidate_everything_for_writes(&mut self) {
        self.for_writes.local_last_op.clear();
        self.for_writes.nonlocal_last_op.clear();
        self.for_writes.invalid_in_parent.all_local = true;
        self.for_writes.invalid_in_parent.all_nonlocal = true;
    }

    fn invalidate_current_scope(&mut self) {
        fn invalidate(this: &mut Ops<'_>) {
            this.local_last_op.clear();
            this.nonlocal_last_op.clear();
        }
        invalidate(&mut self.for_reads);
        invalidate(&mut self.for_writes);
    }

    fn set_last_op(&mut self, target: &'a ir::Ref<ir::Mut>, op: (StmtIndex, WriteOp<'a>)) {
        #[rustfmt::skip]
            let (ops_reads, ops_writes) = if self.refs_used_in_only_one_fn_scope.contains(&target) {
            (&mut self.for_reads.local_last_op, &mut self.for_writes.local_last_op)
        } else {
            (&mut self.for_reads.nonlocal_last_op, &mut self.for_writes.nonlocal_last_op)
        };
        ops_reads.insert(target, op.clone());
        ops_writes.insert(target, op);
        self.for_reads.invalid_in_parent.refs.insert(target);
        self.for_writes.invalid_in_parent.refs.insert(target);
    }

    fn get_last_op<'op>(
        &self,
        ops: &'op Ops<'a>,
        ref_: &&'a ir::Ref<ir::Mut>,
    ) -> Option<&'op (StmtIndex, WriteOp<'a>)> {
        ops.local_last_op
            .get(ref_)
            .or_else(|| ops.nonlocal_last_op.get(ref_))
    }

    fn declare_mut(&mut self, target: &'a ir::Ref<ir::Mut>, val: &'a ir::Ref<ir::Ssa>) {
        let op = (self.cur_index, WriteOp::Declare(val));
        self.set_last_op(target, op);
    }

    fn write_mut(&mut self, target: &'a ir::Ref<ir::Mut>, val: &'a ir::Ref<ir::Ssa>) {
        let op = match self.get_last_op(&self.for_writes, &target) {
            // write -> write (decl)
            Some((declare_index, WriteOp::Declare(_))) => {
                self.mut_ops_to_replace.insert(*declare_index, What::Remove);
                self.mut_ops_to_replace
                    .insert(self.cur_index, What::BecomeDecl);
                (self.cur_index, WriteOp::Declare(val))
            }
            // write -> write
            Some((write_index, WriteOp::Write(_))) => {
                self.mut_ops_to_replace.insert(*write_index, What::Remove);
                (self.cur_index, WriteOp::Write(val))
            }
            None => (self.cur_index, WriteOp::Write(val)),
        };
        self.set_last_op(target, op);
    }

    fn read_mut(&mut self, _target: &'a ir::Ref<ir::Ssa>, source: &'a ir::Ref<ir::Mut>) {
        match self.get_last_op(&self.for_reads, &source) {
            // write -> read
            Some((_, WriteOp::Declare(val))) | Some((_, WriteOp::Write(val))) => {
                self.mut_ops_to_replace
                    .insert(self.cur_index, What::ReadSsa((*val).clone()));
            }
            None => {}
        }
    }
}

impl<'a> Visitor<'a> for CollectLoadStoreInfo<'a> {
    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: &'a ir::Block,
        enter: impl FnOnce(&mut Self, &'a ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            self.refs_used_in_only_one_fn_scope =
                anal::refs::used_in_only_one_fn_scope(&block).collect();
        }

        match ty {
            ScopeTy::Function => {
                // function scopes are analyzed separately
                let mut inner = Self::default();
                mem::swap(
                    &mut inner.refs_used_in_only_one_fn_scope,
                    &mut self.refs_used_in_only_one_fn_scope,
                );
                mem::swap(&mut inner.mut_ops_to_replace, &mut self.mut_ops_to_replace);
                mem::swap(&mut inner.cur_index, &mut self.cur_index);
                let r = enter(&mut inner, block);
                mem::swap(
                    &mut inner.refs_used_in_only_one_fn_scope,
                    &mut self.refs_used_in_only_one_fn_scope,
                );
                mem::swap(&mut inner.mut_ops_to_replace, &mut self.mut_ops_to_replace);
                mem::swap(&mut inner.cur_index, &mut self.cur_index);
                r
            }
            ScopeTy::Normal | ScopeTy::Toplevel => {
                // r->r and w->r can go into scopes, but not w->w (since it might not execute)
                let mut inner = Self::default();
                mem::swap(
                    &mut inner.refs_used_in_only_one_fn_scope,
                    &mut self.refs_used_in_only_one_fn_scope,
                );
                mem::swap(&mut inner.mut_ops_to_replace, &mut self.mut_ops_to_replace);
                mem::swap(&mut inner.cur_index, &mut self.cur_index);
                // todo avoid these clones
                inner.for_reads.local_last_op = self.for_reads.local_last_op.clone();
                inner.for_reads.nonlocal_last_op = self.for_reads.nonlocal_last_op.clone();
                let r = enter(&mut inner, block);
                mem::swap(
                    &mut inner.refs_used_in_only_one_fn_scope,
                    &mut self.refs_used_in_only_one_fn_scope,
                );
                mem::swap(&mut inner.mut_ops_to_replace, &mut self.mut_ops_to_replace);
                mem::swap(&mut inner.cur_index, &mut self.cur_index);
                // invalidate any vars written in the inner scope
                self.invalidate_from_child(inner);
                r
            }
            ScopeTy::Nonlinear => {
                // forwarding can happen across nonlinear scopes, but not into them
                let mut inner = Self::default();
                mem::swap(
                    &mut inner.refs_used_in_only_one_fn_scope,
                    &mut self.refs_used_in_only_one_fn_scope,
                );
                mem::swap(&mut inner.mut_ops_to_replace, &mut self.mut_ops_to_replace);
                mem::swap(&mut inner.cur_index, &mut self.cur_index);
                let r = enter(&mut inner, block);
                mem::swap(
                    &mut inner.refs_used_in_only_one_fn_scope,
                    &mut self.refs_used_in_only_one_fn_scope,
                );
                mem::swap(&mut inner.mut_ops_to_replace, &mut self.mut_ops_to_replace);
                mem::swap(&mut inner.cur_index, &mut self.cur_index);
                // invalidate any vars written in the inner scope
                self.invalidate_from_child(inner);
                r
            }
        }
    }

    fn visit(&mut self, stmt: &'a ir::Stmt) {
        self.cur_index += 1;
        match stmt {
            ir::Stmt::Expr { target, expr } => match expr {
                ir::Expr::ReadMutable { source } => self.read_mut(target, source),
                ir::Expr::Yield { .. } | ir::Expr::Await { .. } | ir::Expr::Call { .. } => {
                    self.invalidate_everything_used_across_fn_scopes()
                }
                ir::Expr::Bool { .. }
                | ir::Expr::Number { .. }
                | ir::Expr::String { .. }
                | ir::Expr::Null
                | ir::Expr::Undefined
                | ir::Expr::This
                | ir::Expr::Read { .. }
                | ir::Expr::ReadGlobal { .. }
                | ir::Expr::ReadMember { .. }
                | ir::Expr::Array { .. }
                | ir::Expr::Object { .. }
                | ir::Expr::RegExp { .. }
                | ir::Expr::Unary { .. }
                | ir::Expr::Binary { .. }
                | ir::Expr::Delete { .. }
                | ir::Expr::Function { .. }
                | ir::Expr::CurrentFunction { .. }
                | ir::Expr::Argument { .. } => {}
            },
            ir::Stmt::DeclareMutable { target, val } => self.declare_mut(target, val),
            ir::Stmt::WriteMutable { target, val } => self.write_mut(target, val),
            ir::Stmt::Return { .. }
            | ir::Stmt::Throw { .. }
            | ir::Stmt::Break { .. }
            | ir::Stmt::Continue { .. } => self.invalidate_everything_for_writes(),
            ir::Stmt::SwitchCase { .. } => self.invalidate_current_scope(),
            ir::Stmt::WriteGlobal { .. }
            | ir::Stmt::WriteMember { .. }
            | ir::Stmt::Debugger { .. }
            | ir::Stmt::Label { .. }
            | ir::Stmt::Loop { .. }
            | ir::Stmt::ForEach { .. }
            | ir::Stmt::IfElse { .. }
            | ir::Stmt::Switch { .. }
            | ir::Stmt::Try { .. } => {}
        }
    }
}

impl Folder for LoadStore {
    type Output = Option<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            let mut collector = CollectLoadStoreInfo::default();
            collector.run_visitor(&block);
            self.mut_ops_to_replace = collector.mut_ops_to_replace;
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        self.cur_index += 1;
        match stmt {
            ir::Stmt::Expr {
                target,
                expr: ir::Expr::ReadMutable { source },
            } => match self.mut_ops_to_replace.remove(&self.cur_index) {
                Some(What::ReadSsa(ssa_ref)) => Some(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Read { source: ssa_ref },
                }),
                Some(What::Remove) | Some(What::BecomeDecl) => {
                    unreachable!("cannot remove/convert mut read to decl")
                }
                None => Some(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::ReadMutable { source },
                }),
            },
            ir::Stmt::DeclareMutable { target, val } => {
                match self.mut_ops_to_replace.remove(&self.cur_index) {
                    Some(What::Remove) => None,
                    Some(What::ReadSsa(_)) | Some(What::BecomeDecl) => {
                        unreachable!("cannot convert mut decl to read or decl")
                    }
                    None => Some(ir::Stmt::DeclareMutable { target, val }),
                }
            }
            ir::Stmt::WriteMutable { target, val } => {
                match self.mut_ops_to_replace.remove(&self.cur_index) {
                    Some(What::Remove) => None,
                    Some(What::BecomeDecl) => Some(ir::Stmt::DeclareMutable { target, val }),
                    Some(What::ReadSsa(_)) => unreachable!("cannot convert mut write to read"),
                    None => Some(ir::Stmt::WriteMutable { target, val }),
                }
            }
            _ => Some(stmt),
        }
    }
}
