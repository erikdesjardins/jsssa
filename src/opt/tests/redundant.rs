use crate::opt::redundant;

case!(
    basic_write_to_read,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let something = 1;
    foo();
    something = x;
    let y = something;
    h = function() { return something; };
"#,
@r###"
_ini = 1
something <= _ini
_fun = <global foo>
<dead> = _fun()
_val = <global x>
something <- _val
<dead> = _val
_ini$1 = _val
y <= _ini$1
_val$1 = <function>:
    _ret = *something
    <return> _ret
<global h> <- _val$1
<dead> = _val$1
"###);

case!(
    basic_write_to_write,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let something = 1;
    foo();
    something = x;
    something = y;
    h = function() { return something; };
"#,
@r###"
_ini = 1
something <= _ini
_fun = <global foo>
<dead> = _fun()
_val = <global x>
<dead> = _val
_val$1 = <global y>
something <- _val$1
<dead> = _val$1
_val$2 = <function>:
    _ret = *something
    <return> _ret
<global h> <- _val$2
<dead> = _val$2
"###);

case!(
    write_to_write_decl,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let something = 1;
    something = x;
    // implicit void init
    var y = 2;
"#,
@r###"
<dead> = <void>
<dead> = 1
_val = <global x>
something <= _val
<dead> = _val
_ini = 2
y <= _ini
"###);

case!(
    forwarded_write_is_redundant,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let something = 1;
    g = something;
    something = 2;
"#,
@r###"
_ini = 1
_val = _ini
<global g> <- _val
<dead> = _val
_val$1 = 2
something <= _val$1
<dead> = _val$1
"###);

case!(
    invalidate_inner_scope_writes,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    let bar = 2;
    let not_written = 2;
    // every ref should be forwarded except the two at the bottom
    if (something) {
        foo;
        foo = 3;
        bar;
        if (something2) {
            bar = 4;
        }
        foo;
        not_written;
    }
    foo; // do not forward
    bar; // do not forward
    not_written;
"#,
@r###"
_ini = 1
foo <= _ini
_ini$1 = 2
bar <= _ini$1
_ini$2 = 2
not_written <= _ini$2
_iff = <global something>
<if> _iff:
    <dead> = _ini
    _val = 3
    foo <- _val
    <dead> = _val
    <dead> = _ini$1
    _iff$1 = <global something2>
    <if> _iff$1:
        _val$1 = 4
        bar <- _val$1
        <dead> = _val$1
    <else>:
        <empty>
    <dead> = _val
    <dead> = _ini$2
<else>:
    <empty>
<dead> = *foo
<dead> = *bar
<dead> = _ini$2
"###);

case!(
    invalidate_inner_scope_writes_for_write,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo;
    foo = 1; // don't drop
    if (something) {
        foo = 2;
    }
    foo = 3;
"#,
@r###"
<dead> = <void>
_val = 1
foo <= _val
<dead> = _val
_iff = <global something>
<if> _iff:
    _val$2 = 2
    foo <- _val$2
    <dead> = _val$2
<else>:
    <empty>
_val$1 = 3
foo <- _val$1
<dead> = _val$1
"###);

case!(
    invalidate_inner_scope_calls,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    while (x)
        invalidate();
    g = foo;
    h = function() { return foo; };
"#,
@r###"
_ini = 1
foo <= _ini
<loop>:
    _whl = <global x>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _fun = <global invalidate>
    <dead> = _fun()
_val = *foo
<global g> <- _val
<dead> = _val
_val$1 = <function>:
    _ret = *foo
    <return> _ret
<global h> <- _val$1
<dead> = _val$1
"###);

case!(
    many_writes,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    foo = 2;
    foo = 3;
    foo = 4;
    foo = 5;
"#,
@r###"
<dead> = 1
_val = 2
<dead> = _val
_val$1 = 3
<dead> = _val$1
_val$2 = 4
<dead> = _val$2
_val$3 = 5
foo <= _val$3
<dead> = _val$3
"###);

case!(
    reads_dont_propagate_to_parent,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    invalidate();
    if (bar) {
        g = foo;
    } else {
        g = foo;
    }
    h = function() { return foo; };
"#,
@r###"
_ini = 1
foo <= _ini
_fun = <global invalidate>
<dead> = _fun()
_iff = <global bar>
<if> _iff:
    _val$1 = *foo
    <global g> <- _val$1
    <dead> = _val$1
<else>:
    _val$1 = *foo
    <global g> <- _val$1
    <dead> = _val$1
_val = <function>:
    _ret = *foo
    <return> _ret
<global h> <- _val
<dead> = _val
"###);

case!(
    switch_invalidate_local,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let outer = 1;
    switch (x) {
        case 1:
            let inner = 2;
            inner;
        case 2:
            inner; // don't forward
    }
    outer;
"#,
@r###"
_ini = 1
outer <= _ini
_swi = <global x>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    _ini$1 = 2
    inner <= _ini$1
    <dead> = _ini$1
    <case> _tst$1:
    <dead> = *inner
<dead> = _ini
"###);

case!(
    across_break_write,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let f;
    while (foo) {
        f += 1; // do not drop
        break;
        f = 3;
    }
"#,
@r###"
_ini = <void>
f <= _ini
<loop>:
    _whl = <global foo>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _lhs = *f
    _rhs = 1
    _val = _lhs + _rhs
    f <- _val
    <dead> = _val
    <break>
    _val$1 = 3
    f <- _val$1
    <dead> = _val$1
"###);

case!(
    across_break_write_nonlocal,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let f;
    while (foo) {
        f += 1; // do not drop
        break;
        f = 3;
    }
    h = function() { return f; }
"#,
@r###"
_ini = <void>
f <= _ini
<loop>:
    _whl = <global foo>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _lhs = *f
    _rhs = 1
    _val$1 = _lhs + _rhs
    f <- _val$1
    <dead> = _val$1
    <break>
    _val$2 = 3
    f <- _val$2
    <dead> = _val$2
_val = <function>:
    _ret = *f
    <return> _ret
<global h> <- _val
<dead> = _val
"###);

case!(
    across_conditional_breaks_write,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let f;
    while (foo) {
        f += 1; // do not drop
        if (bar) {
            break;
        }
        f = 3;
    }
"#,
@r###"
_ini = <void>
f <= _ini
<loop>:
    _whl = <global foo>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _lhs = *f
    _rhs = 1
    _val = _lhs + _rhs
    f <- _val
    <dead> = _val
    _iff = <global bar>
    <if> _iff:
        <break>
    <else>:
        <empty>
    _val$1 = 3
    f <- _val$1
    <dead> = _val$1
"###);

case!(
    across_deep_break_write,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let f;
    outer: while (foo) {
        f += 1; // do not drop
        while (bar) while (bar) {
            break outer;
        }
        f = 3;
    }
"#,
@r###"
_ini = <void>
f <= _ini
<label outer>:
    <loop>:
        _whl = <global foo>
        <if> _whl:
            <empty>
        <else>:
            <break>
        _lhs = *f
        _rhs = 1
        _val = _lhs + _rhs
        f <- _val
        <dead> = _val
        <loop>:
            _whl$1 = <global bar>
            <if> _whl$1:
                <empty>
            <else>:
                <break>
            <loop>:
                _whl$2 = <global bar>
                <if> _whl$2:
                    <empty>
                <else>:
                    <break>
                <break outer>
        _val$1 = 3
        f <- _val$1
        <dead> = _val$1
"###);

case!(
    across_deep_break_write_nonlocal,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let f;
    outer: while (foo) {
        f += 1; // do not drop
        while (bar) while (bar) {
            break outer;
        }
        f = 3;
    }
    h = function() { return f; };
"#,
@r###"
_ini = <void>
f <= _ini
<label outer>:
    <loop>:
        _whl = <global foo>
        <if> _whl:
            <empty>
        <else>:
            <break>
        _lhs = *f
        _rhs = 1
        _val$1 = _lhs + _rhs
        f <- _val$1
        <dead> = _val$1
        <loop>:
            _whl$1 = <global bar>
            <if> _whl$1:
                <empty>
            <else>:
                <break>
            <loop>:
                _whl$2 = <global bar>
                <if> _whl$2:
                    <empty>
                <else>:
                    <break>
                <break outer>
        _val$2 = 3
        f <- _val$2
        <dead> = _val$2
_val = <function>:
    _ret = *f
    <return> _ret
<global h> <- _val
<dead> = _val
"###);

case!(
    across_conditional_breaks_read,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let f;
    while (foo) {
        f += 1;
        if (bar) {
            break;
        }
        f; // forward
    }
"#,
@r###"
_ini = <void>
f <= _ini
<loop>:
    _whl = <global foo>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _lhs = *f
    _rhs = 1
    _val = _lhs + _rhs
    f <- _val
    <dead> = _val
    _iff = <global bar>
    <if> _iff:
        <break>
    <else>:
        <empty>
    <dead> = _val
"###);

case!(
    into_nonlinear_scope,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let f = 1;
    for (;;) {
        f;
        f = 2;
    }
"#,
@r###"
_ini = 1
f <= _ini
<loop>:
    <dead> = *f
    _val = 2
    f <- _val
    <dead> = _val
"###);

case!(
    same_scope_not_invalidated_by_call,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    invalidate();
    if (bar) {
        g = foo;
    }
"#,
@r###"
_ini = 1
foo <= _ini
_fun = <global invalidate>
<dead> = _fun()
_iff = <global bar>
<if> _iff:
    _val = _ini
    <global g> <- _val
    <dead> = _val
<else>:
    <empty>
"###);

case!(
    revalidated_before_entering_scope,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    invalidate();
    foo = 2;
    if (bar) {
        g = foo;
    }
    h = function() { return foo; };
"#,
@r###"
_ini = 1
foo <= _ini
_fun = <global invalidate>
<dead> = _fun()
_val = 2
foo <- _val
<dead> = _val
_iff = <global bar>
<if> _iff:
    _val$2 = _val
    <global g> <- _val$2
    <dead> = _val$2
<else>:
    <empty>
_val$1 = <function>:
    _ret = *foo
    <return> _ret
<global h> <- _val$1
<dead> = _val$1
"###);

case!(
    cross_switch_read_invalidate,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    switch (bar) {
        case 1:
            foo += 1; // do not forward
        case 2:
            g = foo;
    }
"#,
@r###"
_ini = 1
foo <= _ini
_swi = <global bar>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    _lhs = *foo
    _rhs = 1
    _val = _lhs + _rhs
    foo <- _val
    <dead> = _val
    <case> _tst$1:
    _val$1 = *foo
    <global g> <- _val$1
    <dead> = _val$1
"###);

case!(
    cross_switch_write_invalidate,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    switch (bar) {
        case 1:
            foo += 1; // do not drop
        case 2:
            foo = 3;
    }
"#,
@r###"
_ini = 1
foo <= _ini
_swi = <global bar>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    _lhs = *foo
    _rhs = 1
    _val = _lhs + _rhs
    foo <- _val
    <dead> = _val
    <case> _tst$1:
    _val$1 = 3
    foo <- _val$1
    <dead> = _val$1
"###);

case!(
    cross_switch_write_invalidate_nonlocal,
    |cx| cx.run::<redundant::LoadStore>("redundant-load-store"),
    r#"
    let foo = 1;
    switch (bar) {
        case 1:
            foo += 1; // do not drop
        case 2:
            foo = 3;
    }
    h = function() { return foo; }
"#,
@r###"
_ini = 1
foo <= _ini
_swi = <global bar>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    _lhs = *foo
    _rhs = 1
    _val$1 = _lhs + _rhs
    foo <- _val$1
    <dead> = _val$1
    <case> _tst$1:
    _val$2 = 3
    foo <- _val$2
    <dead> = _val$2
_val = <function>:
    _ret = *foo
    <return> _ret
<global h> <- _val
<dead> = _val
"###);
