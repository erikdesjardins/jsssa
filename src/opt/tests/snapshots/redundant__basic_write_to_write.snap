---
created: "2019-04-04T19:26:26.325098100Z"
creator: insta@0.7.1
source: src/opt/tests/redundant.rs
expression: "\n    let something = 1;\n    foo();\n    something = x;\n    something = y;\n"
---
_ini = 1
something <= _ini
_fun = <global foo>
<dead> = _fun()
_val = <global x>
<dead> = _val
_val$1 = <global y>
something <- _val$1
<dead> = _val$1
